realm/realm-library/src/testObjectServer/kotlin/io/realm/BsonTest.kt

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2020 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm
+
+import io.realm.internal.util.BsonConverter
+import org.bson.*
+import org.bson.types.Decimal128
+import org.bson.types.ObjectId
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Ignore
+import org.junit.Test
+import kotlin.test.assertFailsWith
+
+class BsonTest {
+
+    /**
+     * Simple test to verify semantics of org.bson JSON encoding and decoding.
+     */
+    // Only for Bson API evaluation, not testing Realm functionality
+    @Test
+    fun bsonRoundtrip() {
+        val valueInt32   = 42
+        val valueInt64   = 42L
+        val valueString  = "Realm"
+        val valueBoolean = true
+        val valueOid     = ObjectId()
+
+        val document = BsonDocument.parse("{}")
+
+        document.append("arg1", BsonInt32(valueInt32))
+        document.append("arg2", BsonInt64(valueInt64))
+        document.append("arg3", BsonString(valueString))
+        document.append("arg4", BsonBoolean(valueBoolean))
+        document.append("arg5", BsonObjectId(valueOid))
+
+        val roundtrip = BsonDocument.parse(document.toJson())
+        assertEquals(valueInt32,   roundtrip.get("arg1")?.asInt32()?.value)
+        assertEquals(valueInt64,   roundtrip.get("arg2")?.asInt64()?.value)
+        assertEquals(valueString,  roundtrip.get("arg3")?.asString()?.value)
+        assertEquals(valueBoolean, roundtrip.get("arg4")?.asBoolean()?.value)
+        assertEquals(valueOid,     roundtrip.get("arg5")?.asObjectId()?.value)
+
+        // We cannot retrieve bson values differently type, not even if it could fit in the type
+        assertFailsWith<RuntimeException> {
+            roundtrip.getInt32("arg2");
+        }
+        assertFailsWith<RuntimeException> {
+            roundtrip.getInt64("arg1");
+        }
+    }
+
+    /**
+     * Simple test of type conversion between native Java object types and BSON types.
+     */
+    @Test
+    fun bsonConversion() {
```

Comment:
This test seems to cover a lot, both to/from/exceptions. Isn't it possible to split it apart. I can see why to/from are in the same test, but the exception could be moved to its own test

Reply:
Will separate on future updates.